src/sql/workbench/parts/dataExplorer/common/nodeContextUtils.ts

```diff
@@ -0,0 +1,125 @@
+import { IConnectionManagementService } from 'sql/platform/connection/common/connectionManagement';
+import { ConnectionProfile } from 'sql/platform/connection/common/connectionProfile';
+import { ICapabilitiesService } from 'sql/platform/capabilities/common/capabilitiesService';
+
+export class NodeContextUtils extends Disposable {
+
+	static IsDatabaseOrServer = new RawContextKey<boolean>('isDatabaseOrServer', false);
+	static IsMssqlProvided = new RawContextKey<boolean>('isMssqlProvided', false);
+	static isDatabasesFolder = new RawContextKey<boolean>('isDatabasesFolder', false);
+	static IsServer = new RawContextKey<boolean>('isServer', false);
+	static IsWindows = new RawContextKey<boolean>('isWindows', os.platform() === 'win32');
+	static IsCloud = new RawContextKey<boolean>('isCloud', false);
+
+	private isDatabaseOrServerKey: IContextKey<boolean>;
+	private isMssqlProvidedKey: IContextKey<boolean>;
+	private isDatabaseFolderKey: IContextKey<boolean>;
+	private isServerKey: IContextKey<boolean>;
+	private isCloudKey: IContextKey<boolean>;
+
+	constructor(
+		private nodeContextValue: INodeContextValue,
+		@IContextKeyService private contextKeyService: IContextKeyService,
+		@IConnectionManagementService private connectionManagementService: IConnectionManagementService,
+		@ICapabilitiesService private capabilitiesService: ICapabilitiesService
+	) {
+		super();
+		this.bindContextKeys();
+
+		// Set additional node context keys
+		if (this.nodeContextValue.node) {
+			const node = this.nodeContextValue.node;
+			if (node.payload) {
+				this.isDatabasesFolder();
+				this.isProvidedByMssql();
+				this.isCloud();
+			}
+			if (node.type) {
+				this.isServer();
+			}
+			if (node.contextValue) {
+				this.isDatabaseOrServer();
+			}
+		}
+	}
+
+	private bindContextKeys(): void {
+		this.isDatabaseOrServerKey = NodeContextUtils.IsDatabaseOrServer.bindTo(this.contextKeyService);
+		this.isMssqlProvidedKey = NodeContextUtils.IsMssqlProvided.bindTo(this.contextKeyService);
+		this.isDatabaseFolderKey = NodeContextUtils.isDatabasesFolder.bindTo(this.contextKeyService);
+		this.isServerKey = NodeContextUtils.IsServer.bindTo(this.contextKeyService);
+		this.isCloudKey = NodeContextUtils.IsCloud.bindTo(this.contextKeyService);
+	}
+
+	/**
+	 * Helper function to tell whether node is a database or server or not
+	 */
+	private isDatabaseOrServer(): void {
+		if (this.nodeContextValue.node.type) {
+			if (this.nodeContextValue.node.type === NodeType.Database
+				|| this.nodeContextValue.node.type === NodeType.Server) {
+				this.isDatabaseOrServerKey.set(true);
+				return;
+			}
+		} else if (this.nodeContextValue.node.contextValue === NodeType.Database ||
+			this.nodeContextValue.node.contextValue === NodeType.Server) {
+			this.isDatabaseOrServerKey.set(true);
+		}
+	}
+
+	/**
+	 * Helper function to tell whether node provider is MSSQL
+	 */
+	private isProvidedByMssql(): void {
```

Comment:
Why not just have a nodeProvider key that's a string?

Reply:
You can only check for context values with boolean in the Registry

Reply:
In nodeActions.contribution.ts? Or somewhere else?

Anyways that shouldn't be true, you just have to evaluate the expression. An example is the regex we had before :

```
new ContextKeyRegexExpr('viewItem', /.+itemType\.database.*/i))
```

There's also methods like isEqualTo on the ContextKey that you should be able to use like so 

```
NodeContextKey.nodeProvider.IsEqualTo('MSSQL')
```

Reply:
For the item types, I will be using what you have above. But since we only check for `MSSQL` and since this condition is used so often, I'd like to keep it as a boolean.

Reply:
Well...we only check for MSSQL for now. But why wouldn't we make this easily usable for any provider going forward? It seems bad to limit that now for no real gain

---
In reply to: [299617667](https://github.com/microsoft/azuredatastudio/pull/6242#discussion_r299617667) [](ancestors = 299617667)

Reply:
Because it doesn't matter what their provider is?

Reply:
Why do we only care about MSSQL?  Are we going to have the same issue with PGSQL when we add it to Azure Explorer in the next couple months?